e3nn/util/default_type.py

```diff
@@ -0,0 +1,85 @@
+import torch
+
+from functools import wraps
+from ._context import AbstractContextDecoratorManager
+
+
+class torch_default_tensor_type(AbstractContextDecoratorManager):
+    def __init__(self, dtype, device):
+        super().__init__()
+        self.saved_ttype = None
+        self.dtype = dtype
+        self.device = device
+
+    def __enter__(self):
+        if self.dtype is not None or self.device is not None:
+            self.saved_ttype = torch_get_default_tensor_type()
+            torch.set_default_tensor_type(self.ttype)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.saved_ttype is not None:
+            torch.set_default_tensor_type(self.saved_ttype)
+            self.saved_ttype = None
+
+    @property
+    def ttype(self):
+        return torch.empty(0, dtype=self.dtype, device=self.device).type()
+
+
+class torch_default_dtype(AbstractContextDecoratorManager):
+    def __init__(self, dtype):
+        super().__init__()
+        self.saved_dtype = None
+        self.dtype = dtype
+
+    def __enter__(self):
+        if self.dtype is not None:
+            self.saved_dtype = torch.get_default_dtype()
+            torch.set_default_dtype(self.dtype)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.saved_dtype is not None:
+            torch.set_default_dtype(self.saved_dtype)
+            self.saved_dtype = None
+
+
+class torch_default_device(torch_default_tensor_type):
+    def __init__(self, device):
+        super().__init__(None, device)
+
+
+class add_type_kwargs(object):
+    _DOC_NOTE = r"""
+- dtype and device keyword arguments will be passed to torch_default_tensor_type()
+"""
+
+    def __init__(self, dtype=None, device=None):
+        super().__init__()
+        self.dtype = dtype
+        self.device = device
+
+    def __call__(self, f):
+        @wraps(f)
+        def wrapper(*args, dtype=self.dtype, device=self.device, **kwargs):
+            with torch_default_tensor_type(dtype, device):
+                return f(*args, **kwargs)
+
+        if wrapper.__doc__ is not None:
+            if not wrapper.__doc__.endswith("\n"):
+                wrapper.__doc__ += "\n"
+            wrapper.__doc__ += self._DOC_NOTE
+
+        return wrapper
```

Comment:
This class has the bulk of my questions/comments:
1. I'm a little confused why the constructor arguments are needed here. I don't see them being used. You should be able to just supply the parameter defaults as `None` in the `__call__` method.
2. Classes do not need to inherit from `object` anymore, that's an older convention from new/old style classes. You can just write `class MyClass:` and it'll work fine for modern Python versions. Also I don't think you need to call the super constructor for `object`.
3. If you agree that the constructor arguments `dtype` and `device` can be removed, it should be possible to re-write this decorator as a function (not a callable class).
```suggestion
    def __init__(self):
        super().__init__()

    def __call__(self, f):
        @wraps(f)
        def wrapper(*args, dtype=None, device=None, **kwargs):
            with torch_default_tensor_type(dtype, device):
                return f(*args, **kwargs)

        if wrapper.__doc__ is not None:
            if not wrapper.__doc__.endswith("\n"):
                wrapper.__doc__ += "\n"
            wrapper.__doc__ += self._DOC_NOTE

        return wrapper
```

Reply:
> 1. I'm a little confused why the constructor arguments are needed here. I don't see them being used. You should be able to just supply the parameter defaults as `None` in the `__call__` method.
They are not used in this repository yet, but they will make things like this available:

```python3
@add_type_kwargs(dtype=torch.float64)
function_that_will_return_doubletensor_for_default():
    ...

@add_type_kwargs(device="cuda")
function_supposed_to_be_used_on_cuda():
    ...
```
Even these are not being currently used, I think adding such options are better.

> 2. Classes do not need to inherit from `object` anymore, that's an older convention from new/old style classes. You can just write `class MyClass:` and it'll work fine for modern Python versions. Also I don't think you need to call the super constructor for `object`.

You're right, I just added them for the future. Maybe some kind of superclass could be added one day. If this notation is not preferred, then I have no objections removing them.

> 3. If you agree that the constructor arguments `dtype` and `device` can be removed, it should be possible to re-write this decorator as a function (not a callable class).

Actually it was (until jnooree/e3nn@f03a0f2500437740c9e97fe59ccb55d59642256a), but then I decided to add options for flexibility. Please refer the examples above.

Reply:
Thanks @bdice for the review. I also favor simple code. more complex code can be added later, I'm not too worried if the signature of a function needs to change in the future, we have a changelog for that.

Reply:
Removed in jnooree/e3nn@9d0ce0e15ba54a5b0a3bd0e67f72394ec1e9f5fd